packages/xo-server-sdn-controller/src/index.js

```diff
@@ -0,0 +1,443 @@
+                      accessPIF.$network.name_label
+                    }`
+                  )
+                }
+                const starCenterClient = find(this._OVSDBClients, {
+                  host: poolNetwork.starCenter,
+                })
+                if (!starCenterClient) {
+                  log.error(
+                    `Unable to find OVSDB client of star-center host '${await xapi._getOrWaitObject(
+                      poolNetwork.starCenter
+                    ).name_label}'`
+                  )
+                  continue
+                }
+
+                const hostClient = find(this._OVSDBClients, {
+                  host: object.$ref,
+                })
+                if (!hostClient) {
+                  log.error(
+                    `Unable to find OVSDB client of host '${object.name_label}'`
+                  )
+                  continue
+                }
+
+                const network = accessPIF.$network
+                if (newHost) {
+                  await starCenterClient.addInterfaceAndPort(
+                    network.uuid,
+                    network.name_label,
+                    hostClient.address
+                  )
+                }
+              }
+            }
+          }
+        })
+      )
+    })
+
+    objects.on('remove', async objects => {
+      await Promise.all(
+        map(objects, async (object, id) => {
+          // If a Star center host is removed: re-elect a new center where needed
+          const starCenterRef = this._starCenters.get(id)
+          if (starCenterRef) {
+            this._starCenters.delete(id)
+            const poolNetworks = filter(this._poolNetworks, {
+              starCenter: starCenterRef,
+            })
+            let i
+            for (i = 0; i < poolNetworks.length; ++i) {
+              const poolNetwork = poolNetworks[i]
+              const network = await xapi._getOrWaitObject(poolNetwork.network)
+              const newCenter = await this.electNewCenter(network, true)
+              if (newCenter) {
+                poolNetwork.starCenter = newCenter.$ref
+                this._starCenters.set(newCenter.$id, newCenter.$ref)
+              }
+            }
+          }
+
+          // If a network is removed, clean this._poolNetworks from it
+          const networkRef = this._networks.get(id)
+          if (networkRef) {
+            this._networks.delete(id)
+            const poolNetwork = find(this._poolNetworks, {
+              network: networkRef,
+            })
+            if (poolNetwork) {
+              this._poolNetworks.splice(
+                this._poolNetworks.indexOf(poolNetwork),
+                1
+              )
+            }
+          }
+        })
+      )
+    })
+  }
+
+  async setPoolControllerIfNeeded(pool) {
+    if (this.setControllerNotNeeded(pool.$xapi)) {
+      // Nothing to do
+      return
+    }
+
+    this.monitorXapi(pool.$xapi)
+  }
+
+  setControllerNotNeeded(xapi) {
```

Comment:
It's preferable to use positive instead of negative logic.

It's much easier to reason about, especially in case of double negations: `setControllerNeeded()` vs `!setControllerNotNeeded()`.